File: app/src/main/java/za/org/grassroot2/view/adapter/GroupsAdapter.java
package za.org.grassroot2.view.adapter;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.support.v4.graphics.drawable.RoundedBitmapDrawable;
import android.support.v4.graphics.drawable.RoundedBitmapDrawableFactory;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.jakewharton.rxbinding2.view.RxView;
import com.makeramen.roundedimageview.RoundedDrawable;
import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.Observable;
import io.reactivex.subjects.PublishSubject;
import timber.log.Timber;
import za.org.grassroot2.R;
import za.org.grassroot2.model.Group;
import za.org.grassroot2.util.LastModifiedFormatter;

public class GroupsAdapter extends FooterEnabledAdapter<Group> {

    private final Context context;
    private PublishSubject<String> viewClickSubject = PublishSubject.create();

    private PublishSubject<String> groupImageClickSubject = PublishSubject.create();

    public Observable<String> getViewClickObservable() {
        return viewClickSubject;
    }

    public Observable<String> getGroupImageClickObservable() {return groupImageClickSubject; }

    public GroupsAdapter(Context c, List<Group> data) {
        super(data);
        context = c;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder v = super.onCreateViewHolder(parent, viewType);
        if (v == null) {
            final View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_group, parent, false);

            return new GroupViewHolder(view);
        }

        return v;
    }

    @Override
    protected void bindRegularViewHolder(RecyclerView.ViewHolder vh, int position) {
        Group item = items.get(position);
        GroupViewHolder holder = (GroupViewHolder) vh;

        if (item != null) {
            holder.name.setText(item.getName());
            holder.count.setText(holder.count.getContext().getResources().getQuantityString(R.plurals.member_count, item.getMemberCount(), item.getMemberCount()));
            holder.lastModified.setText(LastModifiedFormatter.lastSeen(context, item.getLastTimeChangedServer()));
            holder.organiser.setText("Placeholder");
            holder.letter.setText(item.getName().substring(0, 1));
            holder.letter.setVisibility(View.VISIBLE);

            if(item.getProfileImageUrl() != null){
                Picasso.get()
                        .load(item.getProfileImageUrl())
                        .resizeDimen(R.dimen.profile_photo_width, R.dimen.profile_photo_height)
                        .centerCrop()
                        .into(holder.image);
            }

            RxView.clicks(holder.root)
                    .map(o -> item.getUid())
                    .subscribe(viewClickSubject);

            RxView.clicks(holder.image)
                    .map(object -> item.getUid())
                    .subscribe(groupImageClickSubject);

        }
    }

    static class GroupViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.root)         View     root;
        @BindView(R.id.letter)       TextView letter;
        @BindView(R.id.name)         TextView name;
        @BindView(R.id.organiser)    TextView organiser;
        @BindView(R.id.count)        TextView count;
        @BindView(R.id.lastModified) TextView lastModified;
        @BindView(R.id.image) ImageView image;

        private GroupViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

}

File: app/src/main/java/za/org/grassroot2/view/adapter/GroupsAdapter.kt
package za.org.grassroot2.view.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView

import com.jakewharton.rxbinding2.view.RxView
import com.squareup.picasso.Picasso

import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject
import kotlinx.android.synthetic.main.item_group.view.*
import za.org.grassroot2.R
import za.org.grassroot2.model.Group
import za.org.grassroot2.util.LastModifiedFormatter

class GroupsAdapter(private val context: Context, data: List<Group>) : FooterEnabledAdapter<Group>(data) {
    private val viewClickSubject = PublishSubject.create<String>()


    private val groupImageClickSubject = PublishSubject.create<String>()

    val viewClickObservable: Observable<String>
        get() = viewClickSubject

    val groupImageClickObservable: Observable<String>
        get() = groupImageClickSubject


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder =
            super.createFooterEnabledViewHolder(parent, viewType) ?: nonFooterViewHolder(parent)

    private fun nonFooterViewHolder(parent: ViewGroup): RecyclerView.ViewHolder =
            GroupViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_group, parent, false))

    override fun bindRegularViewHolder(vh: RecyclerView.ViewHolder, position: Int) {
        val item = items[position]
        val holder = vh as GroupViewHolder


        if (item != null) {
            holder.name!!.text = item.name
            holder.count!!.text = holder.count!!.context.resources.getQuantityString(R.plurals.member_count, item.memberCount!!, item.memberCount)
            holder.lastModified!!.text = LastModifiedFormatter.lastSeen(context, item.lastTimeChangedServer)
            holder.organiser!!.text = "Placeholder"
            holder.letter!!.text = item?.name.substring(0, 1)
            holder.letter!!.visibility = View.VISIBLE

            item.profileImageUrl?.let {
                Picasso.get()
                        .load(it)
                        .resizeDimen(R.dimen.profile_photo_width, R.dimen.profile_photo_height)
                        .centerCrop()
                        .into(holder.image)
            }

            RxView.clicks(holder.root!!)
                    .map { o -> item.uid }
                    .subscribe(viewClickSubject)

            RxView.clicks(holder.image!!)
                    .map { `object` -> item.uid }
                    .subscribe(groupImageClickSubject)

        }
    }

    fun setImage(imageUrl: String, groupUid: String) {
        for (group in items) {
            if (group.uid == groupUid) {
                group.profileImageUrl = imageUrl
                notifyItemChanged(items.indexOf(group))
            }
        }
    }

    internal class GroupViewHolder constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {

        var root: View = itemView.root

        var letter: TextView = itemView.letter

        var name: TextView = itemView.name

        var organiser: TextView = itemView.organiser

        var count: TextView = itemView.count

        var lastModified: TextView = itemView.lastModified

        var image: ImageView = itemView.image
    }

}
